### PR TITLE
enable grouping of wheres

### LINKED ISSUE #200: Grouping wheres and or_wheres in parentheses
A feature that would allow turning something like:

``` crystal
Query.where(name: "foo").and { |q| q.where(age: 20).or_where(age: 30) }
```
into:

`SELECT * FROM users WHERE name = "foo" AND (age = 20 OR age = 30)`

### PR DESCRIPTION
**Pull request by jreinert**
- Introduce "where expressions" which can be combined similarily like
  boolean expressions
- Add `#and` and `#or` methods to query for grouping

Fixes #200

### CODE REVIEW

**Comment by jreinert** [comment]
Sorry, this add's quite some complexity to `Query` but I saw no simple way to do this. Feedback very welcome.

**Comment by faultyserver** [comment]
Reading this through I'm not sure I really like the syntax of using nested blocks, but at the same time it makes perfect sense given the context of what's actually happening. Just a subjective thing I guess.

For the complexity in `Query`, I think we're going to be stuck with that until we get a complete arel-like implementation in place, and this is really a good start toward that so I don't mind at all.

Overall it looks good to me! The only addition I would suggest is allowing `and` and `or` to accept another `Query` object and pluck the `where_expression` from it. Something like:

```crystal
nested_where = Query.where(name: "jokke", age: 99)
q = Query.where(foo: "bar").or(nested_where)
```

File: src/crecto/repo/query.cr [hunk 1]
```diff
@@ -406,6 +567,100 @@ module Crecto
...
+      #   Query.where(name: "Bill", age: 20)
+      #   Query.where(name: "Wendy", age: 50)
+      # )
+      #
+      # # => SELECT * FROM users WHERE
+      # #    (city = 'Los Angeles')
+      # #    OR ((name = 'Bill') AND (age = 20))
+      # #    OR ((name = 'Wendy') AND (age = 50))
+      # ```
+      def or(*queries : self)
```
Comment (jreinert) [inline src/crecto/repo/query.cr]: @faultyserver like this

Comment (faultyserver) [reply src/crecto/repo/query.cr]: yeah, like that. I think I'd like it to only accept a single query to be explicit about it, but i don't have a strong opinion there.

Comment (jreinert) [reply src/crecto/repo/query.cr]: Hmm I don't know. Allowing multiple queries makes sense to me.

Comment (jreinert) [reply src/crecto/repo/query.cr]: But maybe adding a method signature for accepting arrays or rather anything responding to reduce could be the logical consequence.

**Pull request merged**